apps/src/music/progress/ProgressManager.ts

```diff
@@ -0,0 +1,129 @@
+// This file contains a generic ProgressManager which any lab can include,
+// if it wants to make progress without reloading the page.
+
+// Abstract class that validates a set of conditions. How
+// the validation works is up to the implementor.
+export abstract class Validator {
+  abstract shouldCheckConditions(): boolean;
+  abstract checkConditions(): void;
+  abstract conditionsMet(conditions: string[]): boolean;
+  abstract clear(): void;
+}
+
+// A validation inside the progression.
+interface Validation {
+  conditions: string[];
+  message: string;
+  next: boolean;
+}
+
+// The definition of the progression.
+interface Progression extends Object {
+  panels: {
+    text: string;
+    toolbox: {
+      [key: string]: string;
+    };
+    sounds: {
+      [key: string]: string;
+    };
+    validations: Validation[];
+  }[];
+}
+
+// The current progress state.
+interface ProgressState {
+  step: number;
+  satisfied: boolean;
+  message?: string | null;
+}
+
+export default class ProgressManager {
+  private progression: Progression;
+  private validator: Validator;
+  private onProgressChange: (state: ProgressState) => void;
```

Comment:
Minor - looking at this again, would it be cleaner to have this be a `onProgressChange: () => void` function that doesn't pass any arguments, and then the listener can call `progressManager.getCurrentState()`? That way we don't have to worry about both updating internal state here, and making sure the right state is dispatched via this callback.